llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,381 @@
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+// return a vector containing virtual registers, that have had a copy to define
+// the virtual register
+SmallVector<Register, 16> getVRegWithCopies(const MachineBasicBlock &MBB) {
+  SmallVector<Register, 16> VRegWithCopies;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      Register Reg = Def.getReg();
+
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
+    }
+  }
+
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  // Add the used physical registers one machine instruction at at time.
+  // This vector is used to determine, if a physical register has already been
+  // defined in the machine basic block.
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+
+  // record the register replacements for debugging purposes.
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
+
+  // Get a list of registers, that are not allowed as a replacement register.
+  // This list gets updated with the newly replaced physical register, so that
+  // this pass does not introduce WAW dependencies.
+  BitVector BlockedPhysRegs = getDefinedPhysRegs(MBB);
+
+  // Collect all the virtual registers that have at least a copy instruction
+  // that defines them. Subregisters may contain constants that may be shared
+  // across different virtual registers. Renaming would reintroduce unnecessary
+  // copies, if physical registers are shared. Also do not rename copies, since
+  // they could be removed in a later pass.
+  SmallVector<Register, 16> VRegWithCopies = getVRegWithCopies(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    // Identity copies will be removed in a later pass, therefore, these are not
+    // real defines of a physical register
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+        continue;
+      if (!isLastDef(MO, MI, *MBB))
+        continue;
+
+      MCPhysReg PreSwapPhysReg = VRM->getPhys(Reg);
+      if (isWorthRenaming(Reg, UsedPhysRegs, VRegWithCopies) &&
+          replaceReg(Reg, BlockedPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
+        ReplacementRegs.push_back(
+            std::make_pair(PreSwapPhysReg, VRM->getPhys(Reg)));
+
+      } else {
+        // keep track of already visited physical registers
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
```

Comment:
Do we also need to extend `UsedPhysRegs` for the `if` case?

Reply:
no, because the `UsedPhysRegs` are used to identify when a physical register is a renaming candidate, by checking when it is being written to twice.

1. I do not reuse replaced registers, so they cannot cause WAW issues with later replacements.
2. During the pass setup, I already blocked all the used physical register of the Basic Block. Therefore I cannot introduce a WAW with my replacement. 

To clarify I added a comment to this section to explain it.